Include/internal/pycore_stackref.h

```diff
@@ -0,0 +1,189 @@
+#ifndef Py_INTERNAL_STACKREF_H
+#define Py_INTERNAL_STACKREF_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#include <stddef.h>
+#include "pycore_gc.h"
+
+// Mark an object as supporting deferred reference counting. This is a no-op
+// in the default (with GIL) build. Objects that use deferred reference
+// counting should be tracked by the GC so that they are eventually collected.
+extern void _PyObject_SetDeferredRefcount(PyObject *op);
+
+static inline int
+_PyObject_HasDeferredRefcount(PyObject *op)
+{
+#ifdef Py_GIL_DISABLED
+    return (op->ob_gc_bits & _PyGC_BITS_DEFERRED) != 0;
+#else
+    return 0;
+#endif
+}
+
+typedef union {
+    uintptr_t bits;
+} _PyStackRef;
+
+static const _PyStackRef Py_STACKREF_NULL = { .bits = 0 };
+
+#ifdef Py_GIL_DISABLED
+    #define Py_TAG_DEFERRED (1)
+    #define Py_TAG (Py_TAG_DEFERRED)
+#else
+    #define Py_TAG 0
+#endif
+
+// Gets a PyObject * from a _PyStackRef
+#if defined(Py_GIL_DISABLED)
+    static inline PyObject *
+    PyStackRef_Get(_PyStackRef tagged) {
+        PyObject *cleared = ((PyObject *)((tagged).bits & (~Py_TAG)));
+        return cleared;
+    }
+#else
+    #define PyStackRef_Get(tagged) ((PyObject *)(uintptr_t)((tagged).bits))
+#endif
+
+// Converts a PyObject * to a PyStackRef, stealing the reference.
+#if defined(Py_GIL_DISABLED)
+    static inline _PyStackRef
+    _PyStackRef_StealRef(PyObject *obj) {
+        // Make sure we don't take an already tagged value.
+        assert(PyStackRef_Get(((_PyStackRef){.bits = ((uintptr_t)(obj))})) == obj);
+        return ((_PyStackRef){.bits = ((uintptr_t)(obj))});
+    }
+    #define PyStackRef_StealRef(obj) _PyStackRef_StealRef(_PyObject_CAST(obj))
+#else
+    #define PyStackRef_StealRef(obj) ((_PyStackRef){.bits = ((uintptr_t)(obj))})
+#endif
+
+#if defined(Py_GIL_DISABLED)
+    static inline PyObject *
+    PyStackRef_StealObject(_PyStackRef tagged) {
+        if ((tagged.bits & Py_TAG_DEFERRED) == Py_TAG_DEFERRED) {
+            assert(_PyObject_HasDeferredRefcount(PyStackRef_Get(tagged)));
+            return Py_NewRef(PyStackRef_Get(tagged));
+        }
+        return PyStackRef_Get(tagged);
+    }
+#else
+    #define PyStackRef_StealObject(tagged) PyStackRef_Get(tagged)
+#endif
+
+static inline void
+_Py_untag_stack_borrowed(PyObject **dst, const _PyStackRef *src, size_t length) {
+    for (size_t i = 0; i < length; i++) {
+        dst[i] = PyStackRef_Get(src[i]);
+    }
+}
+
+static inline void
+_Py_untag_stack_steal(PyObject **dst, const _PyStackRef *src, size_t length) {
+    for (size_t i = 0; i < length; i++) {
+        dst[i] = PyStackRef_StealObject(src[i]);
+    }
+}
+
+
+#define PyStackRef_XSETREF(dst, src) \
+    do { \
+        _PyStackRef *_tmp_dst_ptr = _Py_CAST(_PyStackRef*, &(dst)); \
```

Comment:
I'm skeptical of these `_Py_CAST` operations. They make sense for the corresponding `PyObject*` because we have "subtypes" of `PyObject` that are different C types, but that doesn't exist for `_PyStackRef`